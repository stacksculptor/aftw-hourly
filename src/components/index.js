import AddImages from './AddImages/AddImages';
import AuthorInfo from './AuthorInfo/AuthorInfo';
import BookingInfo from './BookingInfo/BookingInfo';
import Button from './Button/Button';
import CurrencyInput from './CurrencyInput/CurrencyInput';
import DateInput from './DateInput/DateInput';
import Discussion from './Discussion/Discussion';
import FilterPanel from './FilterPanel/FilterPanel';
import HeroSection from './HeroSection/HeroSection';
import Input from './Input/Input';
import LabeledField from './LabeledField/LabeledField';
import ListingCard from './ListingCard/ListingCard';
import ListingCardSmall from './ListingCardSmall/ListingCardSmall';
import LocationAutocompleteInput from './LocationAutocompleteInput/LocationAutocompleteInput';
import Map from './Map/Map';
import MapPanel from './MapPanel/MapPanel';
import Menu from './Menu/Menu';
import MobileFrame from './MobileFrame/MobileFrame';
import ModalInMobile from './ModalInMobile/ModalInMobile';
import NamedLink from './NamedLink/NamedLink';
import NamedRedirect from './NamedRedirect/NamedRedirect';
import OrderDetailsPanel from './OrderDetailsPanel/OrderDetailsPanel';
import OrderDiscussionPanel from './OrderDiscussionPanel/OrderDiscussionPanel';
import PageLayout from './PageLayout/PageLayout';
import PaginationLinks from './PaginationLinks/PaginationLinks';
import Promised from './Promised/Promised';
import RoutesProvider from './RoutesProvider/RoutesProvider';
import SearchResultsPanel from './SearchResultsPanel/SearchResultsPanel';
import StripeBankAccountToken from './StripeBankAccountToken/StripeBankAccountToken';

export {
  AddImages,
  AuthorInfo,
  BookingInfo,
  Button,
  CurrencyInput,
  DateInput,
  Discussion,
  FilterPanel,
  HeroSection,
  Input,
  LabeledField,
  ListingCard,
  ListingCardSmall,
  LocationAutocompleteInput,
  Map,
  MapPanel,
  Menu,
  MobileFrame,
  ModalInMobile,
  NamedLink,
  NamedRedirect,
  OrderDetailsPanel,
  OrderDiscussionPanel,
  PageLayout,
  PaginationLinks,
  Promised,
  RoutesProvider,
  SearchResultsPanel,
  StripeBankAccountToken,
};
