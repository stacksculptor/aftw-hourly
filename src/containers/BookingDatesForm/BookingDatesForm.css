@import '../../marketplace.css';

.root {
  display: flex;
  flex-direction: column;
}

.bookingDates {
  flex-shrink: 0;
  margin-bottom: 46px;

  @media (--viewportMedium) {
    margin-bottom: 64px;
  }
}

.priceBreakdownContainer {
  padding: 0 24px;
  margin-bottom: 40px;

  @media (--viewportMedium) {
    padding: 0;
    margin-bottom: 0;
    margin-top: 7px;
  }
}

.priceBreakdownTitle {
  /* Font */
  color: var(--matterColorAnti);

  margin-top: 5px;
  margin-bottom: 14px;
  @media (--viewportMedium) {
    margin-top: 5px;
    margin-bottom: 25px;
  }
}

.receipt {
  flex-shrink: 0;
  margin: 0 0 24px 0;
}

.error {
  color: var(--failColor);
  margin: 0 24px;
  display: inline-block;
}

.smallPrint {
  @apply --marketplaceTinyFontStyles;
  color: var(--matterColorAnti);
  text-align: center;
  margin: auto 24px 20px 24px;
  flex-shrink: 0;

  @media (--viewportMedium) {
    margin-top: 50px;
    margin-bottom: 22px;
  }
}

.submitButton {
  flex-shrink: 0;
  margin: 0 24px 24px 24px;
  width: calc(100% - 48px);

  @media (--viewportMedium) {
    margin: 0;
    width: 100%;
  }
}
