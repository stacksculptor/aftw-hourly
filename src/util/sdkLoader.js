import * as importedSdk from 'sharetribe-sdk';

let exportSdk;

const isServer = () => typeof window === 'undefined';

if (isServer()) {
  // Use eval to skip webpack from bundling SDK in Node
  // eslint-disable-next-line no-eval
  exportSdk = eval('require')('sharetribe-sdk');
} else {
  exportSdk = importedSdk;
}

const {
  createInstance,
  types,
} = exportSdk;

export { createInstance, types };
