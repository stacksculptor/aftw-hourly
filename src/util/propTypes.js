/**
 * This module defines custom PropTypes shared within the application.
 *
 * To learn about validating React component props with PropTypes, see:
 *
 *     https://facebook.github.io/react/docs/typechecking-with-proptypes.html
 *
 * For component specific PropTypes, it's perfectly ok to inline them
 * to the component itself. If the type is shared or external (SDK or
 * API), however, it should be in this file for sharing with other
 * components.
 *
 * PropTypes should usually be validated only at the lowest level
 * where the props are used, not along the way in parents that pass
 * along the props to their children. Those parents should usually
 * just validate the presense of the prop key and that the value is
 * defined. This way we get the validation errors only in the most
 * specific place and avoid duplicate errros.
 */
import PropTypes from 'prop-types';
import Decimal from 'decimal.js';
import { types as sdkTypes } from './sdkLoader';
import { ensureTransaction } from './data';

const { UUID, LatLng, LatLngBounds, Money } = sdkTypes;
const { arrayOf, bool, func, instanceOf, number, object, oneOf, shape, string } = PropTypes;

// Fixed value
export const value = val => oneOf([val]);

// SDK type instances
export const uuid = instanceOf(UUID);
export const latlng = instanceOf(LatLng);
export const latlngBounds = instanceOf(LatLngBounds);
export const money = instanceOf(Money);

// Configuration for currency formatting
export const currencyConfig = shape({
  style: string.isRequired,
  currency: string.isRequired,
  currencyDisplay: string,
  useGrouping: bool,
  minimumFractionDigits: number,
  maximumFractionDigits: number,
});

// Configuration for a single route
export const route = shape({
  name: string.isRequired,
  path: string.isRequired,
  exact: bool,
  strict: bool,
  component: func.isRequired,
  loadData: func,
});

// Place object from LocationAutocompleteInput
export const place = shape({
  address: string.isRequired,
  origin: latlng.isRequired,
  bounds: latlngBounds, // optional viewport bounds
  country: string, // country code, e.g. FI, US
});

// Denormalised image object
export const image = shape({
  id: uuid.isRequired,
  type: value('image').isRequired,
  attributes: shape({
    sizes: arrayOf(
      shape({
        width: number.isRequired,
        height: number.isRequired,
        name: string.isRequired,
        url: string.isRequired,
      })
    ).isRequired,
  }),
});

// Denormalised user object
export const currentUser = shape({
  id: uuid.isRequired,
  type: value('currentUser').isRequired,
  attributes: shape({
    banned: bool.isRequired,
    email: string.isRequired,
    emailVerified: bool.isRequired,
    profile: shape({
      firstName: string.isRequired,
      lastName: string.isRequired,
      displayName: string.isRequired,
      abbreviatedName: string.isRequired,
      bio: string,
    }).isRequired,
    stripeConnected: bool.isRequired,
  }),
  profileImage: image,
});

// Denormalised user object
export const user = shape({
  id: uuid.isRequired,
  type: value('user').isRequired,
  attributes: shape({
    banned: bool.isRequired,
    profile: shape({
      displayName: string.isRequired,
      abbreviatedName: string.isRequired,
      bio: string,
    }),
  }),
  profileImage: image,
});

// Denormalised listing object
export const listing = shape({
  id: uuid.isRequired,
  type: value('listing').isRequired,
  attributes: shape({
    title: string.isRequired,
    description: string.isRequired,
    address: string.isRequired,
    geolocation: latlng.isRequired,
    closed: bool.isRequired,
    deleted: bool.isRequired,
    price: money,
  }).isRequired,
  author: user,
  images: arrayOf(image),
});

// Denormalised booking object
export const booking = shape({
  id: uuid.isRequired,
  type: value('booking').isRequired,
  attributes: shape({
    end: instanceOf(Date).isRequired,
    start: instanceOf(Date).isRequired,
  }),
});

// When the customer requests a booking, a transaction is created. The
// initial state is preauthorized that is transitioned with the
// initial preauthorize transition. The customer can see this
// transaction in the OrderPage that is linked from the InboxPage. The
// provider sees the transaction in the SalePage.
export const TX_TRANSITION_PREAUTHORIZE = 'transition/preauthorize';

// When the provider accepts or declines a transaction from the
// SalePage, it is transitioned with the accept or decline transition.
export const TX_TRANSITION_ACCEPT = 'transition/accept';
export const TX_TRANSITION_DECLINE = 'transition/decline';

// If the backend automatically declines the transaction, it is
// transitioned with the auto-decline transition.
export const TX_TRANSITION_AUTO_DECLINE = 'transition/auto-decline';

// Admin can also cancel the transition.
export const TX_TRANSITION_CANCEL = 'transition/cancel';

// If the is marked as delivered in the backend, it is transitioned
// with the mark-delivered transition.
export const TX_TRANSITION_MARK_DELIVERED = 'transition/mark-delivered';

// Review transitions
export const TX_TRANSITION_REVIEW_BY_PROVIDER_FIRST = 'transition/review-by-provider-first';
export const TX_TRANSITION_REVIEW_BY_PROVIDER_SECOND = 'transition/review-by-provider-second';
export const TX_TRANSITION_REVIEW_BY_CUSTOMER_FIRST = 'transition/review-by-customer-first';
export const TX_TRANSITION_REVIEW_BY_CUSTOMER_SECOND = 'transition/review-by-customer-second';
export const TX_TRANSITION_MARK_REVIEWED_BY_CUSTOMER = 'transition/mark-reviewed-by-customer';
export const TX_TRANSITION_MARK_REVIEWED_BY_PROVIDER = 'transition/mark-reviewed-by-provider';
export const TX_TRANSITION_AUTO_COMPLETE_WITHOUT_REVIEWS =
  'transition/auto-complete-without-reviews';

export const TX_TRANSITIONS = [
  TX_TRANSITION_PREAUTHORIZE,
  TX_TRANSITION_ACCEPT,
  TX_TRANSITION_DECLINE,
  TX_TRANSITION_AUTO_DECLINE,
  TX_TRANSITION_CANCEL,
  TX_TRANSITION_MARK_DELIVERED,
  TX_TRANSITION_REVIEW_BY_PROVIDER_FIRST,
  TX_TRANSITION_REVIEW_BY_PROVIDER_SECOND,
  TX_TRANSITION_REVIEW_BY_CUSTOMER_FIRST,
  TX_TRANSITION_REVIEW_BY_CUSTOMER_SECOND,
  TX_TRANSITION_MARK_REVIEWED_BY_CUSTOMER,
  TX_TRANSITION_MARK_REVIEWED_BY_PROVIDER,
  TX_TRANSITION_AUTO_COMPLETE_WITHOUT_REVIEWS,
];

// Roles of actors that perform transaction transitions
export const TX_TRANSITION_ACTOR_CUSTOMER = 'customer';
export const TX_TRANSITION_ACTOR_PROVIDER = 'provider';
export const TX_TRANSITION_ACTOR_SYSTEM = 'system';
export const TX_TRANSITION_ACTOR_OPERATOR = 'operator';

export const TX_TRANSITION_ACTORS = [
  TX_TRANSITION_ACTOR_CUSTOMER,
  TX_TRANSITION_ACTOR_PROVIDER,
  TX_TRANSITION_ACTOR_SYSTEM,
  TX_TRANSITION_ACTOR_OPERATOR,
];

const txLastTransition = tx => ensureTransaction(tx).attributes.lastTransition;

export const txIsPreauthorized = tx => txLastTransition(tx) === TX_TRANSITION_PREAUTHORIZE;

export const txIsAccepted = tx => txLastTransition(tx) === TX_TRANSITION_ACCEPT;

export const txIsDeclined = tx => txLastTransition(tx) === TX_TRANSITION_DECLINE;

export const txIsAutodeclined = tx => txLastTransition(tx) === TX_TRANSITION_AUTO_DECLINE;

export const txIsDeclinedOrAutodeclined = tx => txIsDeclined(tx) || txIsAutodeclined(tx);

export const txIsCanceled = tx => txLastTransition(tx) === TX_TRANSITION_CANCEL;

export const txIsDelivered = tx => txLastTransition(tx) === TX_TRANSITION_MARK_DELIVERED;

export const txTransition = shape({
  at: instanceOf(Date).isRequired,
  by: oneOf(TX_TRANSITION_ACTORS).isRequired,
  transition: oneOf(TX_TRANSITIONS).isRequired,
});

// Check if tx transition is followed by a state where
// reviews are completed
export const areReviewsCompleted = transition => {
  return [
    TX_TRANSITION_REVIEW_BY_PROVIDER_SECOND,
    TX_TRANSITION_REVIEW_BY_CUSTOMER_SECOND,
    TX_TRANSITION_MARK_REVIEWED_BY_CUSTOMER,
    TX_TRANSITION_MARK_REVIEWED_BY_PROVIDER,
    TX_TRANSITION_AUTO_COMPLETE_WITHOUT_REVIEWS,
  ].includes(transition);
};

// Check if a user giving a review is related to
// given tx transition.
export const isReviewTransition = transition => {
  return [
    TX_TRANSITION_REVIEW_BY_PROVIDER_FIRST,
    TX_TRANSITION_REVIEW_BY_CUSTOMER_FIRST,
    TX_TRANSITION_REVIEW_BY_PROVIDER_SECOND,
    TX_TRANSITION_REVIEW_BY_CUSTOMER_SECOND,
  ].includes(transition);
};

// Possible amount of stars in a review
export const REVIEW_RATINGS = [1, 2, 3, 4, 5];

// A review on a user
export const review = shape({
  id: uuid.isRequired,
  attributes: shape({
    at: instanceOf(Date).isRequired,
    content: string,
    rating: oneOf(REVIEW_RATINGS),
    state: string.isRequired,
    type: string.isRequired,
  }).isRequired,
  author: user,
  subject: user,
});

// Denormalised transaction object
export const transaction = shape({
  id: uuid.isRequired,
  type: value('transaction').isRequired,
  attributes: shape({
    createdAt: instanceOf(Date).isRequired,
    lastTransitionedAt: instanceOf(Date).isRequired,
    lastTransition: oneOf(TX_TRANSITIONS).isRequired,
    payinTotal: money.isRequired,
    payoutTotal: money.isRequired,
    lineItems: arrayOf(
      shape({
        code: string.isRequired,
        quantity: instanceOf(Decimal),
        unitPrice: money.isRequired,
        lineTotal: money.isRequired,
        reversal: bool.isRequired,
      })
    ).isRequired,
    transitions: arrayOf(txTransition).isRequired,
  }),
  booking,
  listing,
  customer: user,
  provider: user,
  reviews: arrayOf(review),
});

// Denormalised transaction message
export const message = shape({
  id: uuid.isRequired,
  type: value('message').isRequired,
  attributes: shape({
    at: instanceOf(Date).isRequired,
    content: string.isRequired,
  }).isRequired,
  sender: user,
});

// Pagination information in the response meta
export const pagination = shape({
  page: number.isRequired,
  perPage: number.isRequired,
  totalItems: number.isRequired,
  totalPages: number.isRequired,
});

export const ERROR_CODE_TRANSACTION_LISTING_NOT_FOUND = 'transaction-listing-not-found';
export const ERROR_CODE_PAYMENT_FAILED = 'transaction-payment-failed';
export const ERROR_CODE_EMAIL_TAKEN = 'email-taken';
export const ERROR_CODE_EMAIL_NOT_FOUND = 'email-not-found';
export const ERROR_CODE_EMAIL_NOT_VERIFIED = 'email-unverified';
export const ERROR_CODE_TOO_MANY_VERIFICATION_REQUESTS = 'email-too-many-verification-requests';
export const ERROR_CODE_UPLOAD_OVER_LIMIT = 'request-upload-over-limit';
const ERROR_CODES = [
  ERROR_CODE_TRANSACTION_LISTING_NOT_FOUND,
  ERROR_CODE_PAYMENT_FAILED,
  ERROR_CODE_EMAIL_TAKEN,
  ERROR_CODE_EMAIL_NOT_FOUND,
  ERROR_CODE_EMAIL_NOT_VERIFIED,
  ERROR_CODE_TOO_MANY_VERIFICATION_REQUESTS,
  ERROR_CODE_UPLOAD_OVER_LIMIT,
];

// API error
// TODO this is likely to change soonish
export const apiError = shape({
  id: uuid.isRequired,
  status: number.isRequired,
  code: oneOf(ERROR_CODES).isRequired,
  title: string.isRequired,
  details: shape({
    data: object,
    msg: string,
  }),
});

// Storable error prop type. (Error object should not be stored as it is.)
export const error = shape({
  type: value('error').isRequired,
  name: string.isRequired,
  message: string,
  status: number,
  statusText: string,
  apiErrors: arrayOf(apiError),
});
