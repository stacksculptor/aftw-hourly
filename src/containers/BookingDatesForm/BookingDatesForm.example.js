/* eslint-disable no-console */
import { types } from '../../util/sdkLoader';
import BookingDatesForm from './BookingDatesForm';

export const Empty = {
  component: BookingDatesForm,
  props: {
    onSubmit: values => {
      console.log('Submit BookingDatesForm with values:', values);
    },
    price: new types.Money(1099, 'USD'),
  },
  group: 'forms',
};
