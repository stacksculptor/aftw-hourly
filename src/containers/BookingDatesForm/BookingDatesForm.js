import React, { PropTypes } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Field, reduxForm, formValueSelector, propTypes as formPropTypes } from 'redux-form';
import { FormattedMessage, intlShape, injectIntl } from 'react-intl';
import { isInclusivelyAfterDay, isInclusivelyBeforeDay } from 'react-dates';
import moment from 'moment';
import { types } from '../../util/sdkLoader';
import { required } from '../../util/validators';
import { Button, BookingInfo, DateInput } from '../../components';
import css from './BookingDatesForm.css';

const EnhancedDateInput = props => {
  const { input, isOutsideRange, labelMessage, placeholder, meta } = props;
  const { onBlur, onChange, onFocus, value } = input;
  const { touched, error } = meta;
  const maybeIsOutsideRange = isOutsideRange ? { isOutsideRange } : {};
  const inputProps = { ...maybeIsOutsideRange, onBlur, onChange, onFocus, placeholder, value };

  return (
    <div>
      {labelMessage ? <label htmlFor="bookingStart">{labelMessage}</label> : null}
      <DateInput {...inputProps} />
      {touched && error ? <span className={css.error}>{error}</span> : null}
    </div>
  );
};

EnhancedDateInput.defaultProps = {
  input: null,
  isOutsideRange: null,
  labelMessage: null,
  placeholder: 'Date',
};

const { bool, func, instanceOf, node, object, shape, string } = PropTypes;

EnhancedDateInput.propTypes = {
  input: object,
  isOutsideRange: func,
  labelMessage: node,
  meta: shape({
    touched: bool,
    error: string,
  }).isRequired,
  placeholder: string,
};

export const BookingDatesFormComponent = props => {
  const {
    bookingStart,
    bookingEnd,
    className,
    handleSubmit,
    intl,
    price,
    pristine,
    submitting,
  } = props;
  const placeholderText = intl.formatMessage({ id: 'BookingDatesForm.placeholder' });
  const bookingStartLabel = intl.formatMessage({ id: 'BookingDatesForm.bookingStartTitle'});
  const bookingEndLabel = intl.formatMessage({ id: 'BookingDatesForm.bookingEndTitle'});
  const requiredMessage = intl.formatMessage({ id: 'BookingDatesForm.requiredDate' });

  // A day is outside range if it is between today and booking end date (if end date has been chosen)
  const isOutsideRangeStart = bookingEnd
    ? {
        isOutsideRange: day =>
          !(isInclusivelyAfterDay(day, moment()) &&
            isInclusivelyBeforeDay(day, moment(bookingEnd))),
      }
    : {};

  // A day is outside range if it is after booking start date (or today if none is chosen)
  const startOfBookingEndRange = bookingStart ? moment(bookingStart) : moment();
  const isOutsideRangeEnd = bookingStart
    ? { isOutsideRange: day => !isInclusivelyAfterDay(day, startOfBookingEndRange) }
    : {};

  const bookingInfo = price
    ? <BookingInfo
        className={css.receipt}
        bookingStart={bookingStart}
        bookingEnd={bookingEnd}
        unitPrice={price}
      />
    : null;

  const invalid = !(bookingStart && bookingEnd);

  return (
    <form className={className} onSubmit={handleSubmit}>
      <Field
        name="bookingStart"
        labelMessage={bookingStartLabel}
        component={EnhancedDateInput}
        format={null}
        placeholder={placeholderText}
        {...isOutsideRangeStart}
        validate={[required(requiredMessage)]}
      />
      <Field
        name="bookingEnd"
        labelMessage={bookingEndLabel}
        component={EnhancedDateInput}
        format={null}
        placeholder={placeholderText}
        {...isOutsideRangeEnd}
        validate={[required(requiredMessage)]}
      />
      {bookingInfo}
      <p className={css.smallPrint}>
        <FormattedMessage id="BookingDatesForm.youWontBeChargedInfo" />
      </p>
      <Button type="submit" disabled={pristine || submitting || invalid}>
        <FormattedMessage id="BookingDatesForm.requestToBook" />
      </Button>
    </form>
  );
};

BookingDatesFormComponent.propTypes = {
  ...formPropTypes,
  intl: intlShape.isRequired,
  price: instanceOf(types.Money).isRequired,
};

const formName = 'bookingDates';

// When a field depends on the value of another field, we must connect
// to the store and select the required values to inject to the
// component.
//
// See: http://redux-form.com/6.6.1/examples/selectingFormValues/
const selector = formValueSelector(formName);
const mapStateToProps = state => {
  return selector(state, 'bookingStart', 'bookingEnd');
};

const BookingDatesForm = compose(
  connect(mapStateToProps),
  reduxForm({ form: formName }),
  injectIntl
)(BookingDatesFormComponent);
BookingDatesForm.displayName = 'BookingDatesForm';

export default BookingDatesForm;
