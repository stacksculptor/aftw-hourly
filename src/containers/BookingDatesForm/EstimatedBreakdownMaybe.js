import moment from 'moment';
import Decimal from 'decimal.js';
import { types as sdkTypes } from '../../util/sdkLoader';
import { dateFromLocalToAPI, nightsBetween, daysBetween } from '../../util/dates';
import {
  LINE_ITEM_DAY,
  LINE_ITEM_NIGHT,
  LINE_ITEM_UNITS,
  TRANSITION_REQUEST,
  TX_TRANSITION_ACTOR_CUSTOMER,
} from '../../util/types';
import { unitDivisor, convertMoneyToNumber, convertUnitToSubUnit } from '../../util/currency';
import { BookingBreakdown } from '../../components';

import css from './BookingDatesForm.css';

const { Money, UUID } = sdkTypes;

const estimatedTotalPrice = (unitPrice, unitCount) => {
  const numericPrice = convertMoneyToNumber(unitPrice);
  const numericTotalPrice = new Decimal(numericPrice).times(unitCount).toNumber();
  return new Money(
    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
    unitPrice.currency
  );
};

// When we cannot speculatively initiate a transaction (i.e. logged
// out), we must estimate the booking breakdown. This function creates
// an estimated transaction object for that use case.
const estimatedTransaction = (unitType, bookingStart, bookingEnd, unitPrice, quantity) => {
  const now = new Date();
  const isNightly = unitType === LINE_ITEM_NIGHT;
  const isDaily = unitType === LINE_ITEM_DAY;

  const unitCount = isNightly
    ? nightsBetween(bookingStart, bookingEnd)
    : isDaily ? daysBetween(bookingStart, bookingEnd) : quantity;

  const totalPrice = estimatedTotalPrice(unitPrice, unitCount);

  // bookingStart: "Fri Mar 30 2018 12:00:00 GMT-1100 (SST)" aka "Fri Mar 30 2018 23:00:00 GMT+0000 (UTC)"
  // Server normalizes night/day bookings to start from 00:00 UTC aka "Thu Mar 29 2018 13:00:00 GMT-1100 (SST)"
  // The result is: local timestamp.subtract(12h).add(timezoneoffset) (in eg. -23 h)

  // local noon -> startOf('day') => 00:00 local => remove timezoneoffset => 00:00 API (UTC)
  const serverDayStart = dateFromLocalToAPI(
    moment(bookingStart)
      .startOf('day')
      .toDate()
  );
  const serverDayEnd = dateFromLocalToAPI(
    moment(bookingEnd)
      .startOf('day')
      .toDate()
  );

  return {
    id: new UUID('estimated-transaction'),
    type: 'transaction',
    attributes: {
      createdAt: now,
      lastTransitionedAt: now,
      lastTransition: TRANSITION_REQUEST,
      payinTotal: totalPrice,
      payoutTotal: totalPrice,
      lineItems: [
        {
          code: unitType,
          includeFor: ['customer', 'provider'],
          unitPrice: unitPrice,
          quantity: new Decimal(unitCount),
          lineTotal: totalPrice,
          reversal: false,
        },
      ],
      transitions: [
        {
          at: now,
          by: TX_TRANSITION_ACTOR_CUSTOMER,
          transition: TRANSITION_REQUEST,
        },
      ],
    },
    booking: {
      id: new UUID('estimated-booking'),
      type: 'booking',
      attributes: {
        start: serverDayStart,
        end: serverDayEnd,
      },
    },
  };
};

const estimatedBreakdown = (unitType, bookingStart, bookingEnd, unitPrice, quantity) => {
  const isUnits = unitType === LINE_ITEM_UNITS;
  const quantityIfUsingUnits = !isUnits || Number.isInteger(quantity);
  const canEstimatePrice = bookingStart && bookingEnd && unitPrice && quantityIfUsingUnits;
  if (!canEstimatePrice) {
    return null;
  }

  const tx = estimatedTransaction(unitType, bookingStart, bookingEnd, unitPrice, quantity);

  return (
    <BookingBreakdown
      className={css.receipt}
      userRole="customer"
      unitType={unitType}
      transaction={tx}
      booking={tx.booking}
    />
  );
};
