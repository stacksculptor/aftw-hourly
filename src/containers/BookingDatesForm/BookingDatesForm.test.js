import React from 'react';
import { shallow } from 'enzyme';
import Decimal from 'decimal.js';
import { types } from '../../util/sdkLoader';
import { renderShallow } from '../../util/test-helpers';
import { fakeIntl, fakeFormProps } from '../../util/test-data';
import * as propTypes from '../../util/propTypes';
import { BookingBreakdown } from '../../components';
import { BookingDatesFormComponent } from './BookingDatesForm';

const { Money } = types;

const noop = () => null;

describe('BookingDatesForm', () => {
  it('matches snapshot without selected dates', () => {
    const tree = renderShallow(
      <BookingDatesFormComponent
        {...fakeFormProps}
        unitType={propTypes.LINE_ITEM_NIGHT}
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        price={new Money(1099, 'USD')}
        bookingDates={{}}
        startDatePlaceholder="today"
        endDatePlaceholder="tomorrow"
      />
    );
    expect(tree).toMatchSnapshot();
  });
  it('matches selected dates', () => {
    const price = new Money(1099, 'USD');
    const startDate = new Date(Date.UTC(2017, 3, 14));
    const endDate = new Date(Date.UTC(2017, 3, 16));
    const tree = shallow(
      <BookingDatesFormComponent
        {...fakeFormProps}
        unitType={propTypes.LINE_ITEM_NIGHT}
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        price={price}
        bookingDates={{ startDate, endDate }}
        startDatePlaceholder="today"
        endDatePlaceholder="tomorrow"
      />
    );
    const breakdown = tree.find(BookingBreakdown);
    const { userRole, transaction, booking } = breakdown.props();
    expect(userRole).toEqual('customer');
    expect(booking.attributes.start).toEqual(startDate);
    expect(booking.attributes.end).toEqual(endDate);
    expect(transaction.attributes.lastTransition).toEqual(propTypes.TX_TRANSITION_PREAUTHORIZE);
    expect(transaction.attributes.payinTotal).toEqual(new Money(2198, 'USD'));
    expect(transaction.attributes.payoutTotal).toEqual(new Money(2198, 'USD'));
    expect(transaction.attributes.lineItems).toEqual([
      {
        code: 'line-item/night',
        includeFor: ['customer', 'provider'],
        unitPrice: price,
        quantity: new Decimal(2),
        lineTotal: new Money(2198, 'USD'),
        reversal: false,
      },
    ]);
  });
});
