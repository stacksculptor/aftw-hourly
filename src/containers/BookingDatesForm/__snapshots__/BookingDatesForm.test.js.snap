exports[`BookingDatesForm matches snapshot with selected dates 1`] = `
<form
  className={null}
  onSubmit={[Function]}>
  <DateRangeInputField
    endDateId="fakeTestForm.bookingEndDate"
    endDateLabel="BookingDatesForm.bookingEndTitle"
    endDatePlaceholderText="tomorrow"
    format={null}
    name="bookingDates"
    startDateId="fakeTestForm.bookingStartDate"
    startDateLabel="BookingDatesForm.bookingStartTitle"
    startDatePlaceholderText="today"
    useMobileMargins={true}
    validate={
      Array [
        [Function],
        [Function],
      ]
    } />
  <div>
    <h3>
      <FormattedMessage
        id="BookingDatesForm.priceBreakdownTitle"
        values={Object {}} />
    </h3>
    <BookingBreakdown
      bookingEnd={2017-04-16T00:00:00.000Z}
      bookingStart={2017-04-14T00:00:00.000Z}
      lineItems={
        Array [
          Object {
            "code": "line-item/night",
            "lineTotal": Money {
              "amount": 2198,
              "currency": "USD",
            },
            "quantity": "2",
            "unitPrice": Money {
              "amount": 1099,
              "currency": "USD",
            },
          },
        ]
      }
      payinTotal={
        Money {
          "amount": 2198,
          "currency": "USD",
        }
      }
      unitPrice={
        Money {
          "amount": 1099,
          "currency": "USD",
        }
      }
      userRole="customer" />
  </div>
  <p>
    <FormattedMessage
      id="BookingDatesForm.youWontBeChargedInfo"
      values={Object {}} />
  </p>
  <PrimaryButton
    disabled={false}
    type="submit">
    <FormattedMessage
      id="BookingDatesForm.requestToBook"
      values={Object {}} />
  </PrimaryButton>
</form>
`;

exports[`BookingDatesForm matches snapshot without selected dates 1`] = `
<form
  className={null}
  onSubmit={[Function]}>
  <DateRangeInputField
    endDateId="fakeTestForm.bookingEndDate"
    endDateLabel="BookingDatesForm.bookingEndTitle"
    endDatePlaceholderText="tomorrow"
    format={null}
    name="bookingDates"
    startDateId="fakeTestForm.bookingStartDate"
    startDateLabel="BookingDatesForm.bookingStartTitle"
    startDatePlaceholderText="today"
    useMobileMargins={true}
    validate={
      Array [
        [Function],
        [Function],
      ]
    } />
  <p>
    <FormattedMessage
      id="BookingDatesForm.youWontBeChargedInfo"
      values={Object {}} />
  </p>
  <PrimaryButton
    disabled={true}
    type="submit">
    <FormattedMessage
      id="BookingDatesForm.requestToBook"
      values={Object {}} />
  </PrimaryButton>
</form>
`;
